File: frontend/android-app/app/src/test/java/com/codenized/planixor/data/sync/NotificationPurgeServiceTest.kt
package com.codenized.planixor.data.sync

import com.codenized.planixor.data.local.CalendarEventDao
import com.codenized.planixor.data.local.CalendarEventEntity
import com.codenized.planixor.data.local.NotificationRecordDao
import com.codenized.planixor.data.local.NotificationRecordEntity
import io.mockk.coEvery
import io.mockk.coVerify
import io.mockk.mockk
import kotlinx.coroutines.test.runTest
import org.junit.Assert.assertEquals
import org.junit.Before
import org.junit.Test
import java.time.LocalDate
import java.time.format.DateTimeFormatter

/**
 * Unit tests for NotificationPurgeService.
 *
 * Property 3: Client purge identifies correct records
 * Validates: Requirements 2.1, 2.3, 2.7
 */
class NotificationPurgeServiceTest {

    private lateinit var notificationRecordDao: NotificationRecordDao
    private lateinit var calendarEventDao: CalendarEventDao
    private lateinit var purgeService: NotificationPurgeService

    private val today = LocalDate.now().format(DateTimeFormatter.ISO_LOCAL_DATE)
    private val yesterday = LocalDate.now().minusDays(1).format(DateTimeFormatter.ISO_LOCAL_DATE)
    private val tomorrow = LocalDate.now().plusDays(1).format(DateTimeFormatter.ISO_LOCAL_DATE)

    @Before
    fun setup() {
        notificationRecordDao = mockk(relaxed = true)
        calendarEventDao = mockk(relaxed = true)
        purgeService = NotificationPurgeService(notificationRecordDao, calendarEventDao)
    }

    @Test
    fun `should return 0 when no notification records exist`() = runTest {
        coEvery { notificationRecordDao.getAll() } returns emptyList()
        coEvery { calendarEventDao.getAll() } returns emptyList()

        val result = purgeService.purgePastNotifications()

        assertEquals(0, result)
        coVerify(exactly = 0) { notificationRecordDao.deleteByIds(any()) }
    }

    @Test
    fun `should purge records whose calendarEvent startDay is before today`() = runTest {
        val pastEventId = "event-past-1"
        val notificationId = "notif-1"

        coEvery { notificationRecordDao.getAll() } returns listOf(
            createNotificationRecord(id = notificationId, calendarEventId = pastEventId),
        )
        coEvery { calendarEventDao.getAll() } returns listOf(
            createCalendarEvent(id = pastEventId, startDay = yesterday),
        )

        val result = purgeService.purgePastNotifications()

        assertEquals(1, result)
        coVerify { notificationRecordDao.deleteByIds(listOf(notificationId)) }
    }

    @Test
    fun `should NOT purge records whose calendarEvent startDay equals today`() = runTest {
        val todayEventId = "event-today-1"
        val notificationId = "notif-1"

        coEvery { notificationRecordDao.getAll() } returns listOf(
            createNotificationRecord(id = notificationId, calendarEventId = todayEventId),
        )
        coEvery { calendarEventDao.getAll() } returns listOf(
            createCalendarEvent(id = todayEventId, startDay = today),
        )

        val result = purgeService.purgePastNotifications()

        assertEquals(0, result)
        coVerify(exactly = 0) { notificationRecordDao.deleteByIds(any()) }
    }

    @Test
    fun `should NOT purge records whose calendarEvent startDay is after today`() = runTest {
        val futureEventId = "event-future-1"
        val notificationId = "notif-1"

        coEvery { notificationRecordDao.getAll() } returns listOf(
            createNotificationRecord(id = notificationId, calendarEventId = futureEventId),
        )
        coEvery { calendarEventDao.getAll() } returns listOf(
            createCalendarEvent(id = futureEventId, startDay = tomorrow),
        )

        val result = purgeService.purgePastNotifications()

        assertEquals(0, result)
        coVerify(exactly = 0) { notificationRecordDao.deleteByIds(any()) }
    }

    @Test
    fun `should purge orphaned records with no matching calendarEvent`() = runTest {
        val orphanedNotificationId = "notif-orphaned-1"
        val nonExistentEventId = "event-does-not-exist"

        coEvery { notificationRecordDao.getAll() } returns listOf(
            createNotificationRecord(id = orphanedNotificationId, calendarEventId = nonExistentEventId),
        )
        coEvery { calendarEventDao.getAll() } returns emptyList()

        val result = purgeService.purgePastNotifications()

        assertEquals(1, result)
        coVerify { notificationRecordDao.deleteByIds(listOf(orphanedNotificationId)) }
    }

    @Test
    fun `should handle mixed records correctly`() = runTest {
        val pastEventId = "event-past"
        val todayEventId = "event-today"
        val futureEventId = "event-future"
        val orphanedEventId = "event-orphan"

        val pastNotifId = "notif-past"
        val todayNotifId = "notif-today"
        val futureNotifId = "notif-future"
        val orphanedNotifId = "notif-orphaned"

        coEvery { notificationRecordDao.getAll() } returns listOf(
            createNotificationRecord(id = pastNotifId, calendarEventId = pastEventId),
            createNotificationRecord(id = todayNotifId, calendarEventId = todayEventId),
            createNotificationRecord(id = futureNotifId, calendarEventId = futureEventId),
            createNotificationRecord(id = orphanedNotifId, calendarEventId = orphanedEventId),
        )
        coEvery { calendarEventDao.getAll() } returns listOf(
            createCalendarEvent(id = pastEventId, startDay = yesterday),
            createCalendarEvent(id = todayEventId, startDay = today),
            createCalendarEvent(id = futureEventId, startDay = tomorrow),
        )

        val result = purgeService.purgePastNotifications()

        assertEquals(2, result)
        coVerify {
            notificationRecordDao.deleteByIds(
                match { ids ->
                    ids.containsAll(listOf(pastNotifId, orphanedNotifId)) && ids.size == 2
                },
            )
        }
    }

    @Test
    fun `should return 0 when DAO operation fails and not throw`() = runTest {
        coEvery { notificationRecordDao.getAll() } throws RuntimeException("DB error")

        val result = purgeService.purgePastNotifications()

        assertEquals(0, result)
    }

    // --- Helper functions ---

    private fun createNotificationRecord(
        id: String,
        calendarEventId: String,
    ): NotificationRecordEntity = NotificationRecordEntity(
        id = id,
        calendarEventId = calendarEventId,
        alertOffset = 10,
        triggerTime = System.currentTimeMillis(),
        isDelivered = false,
        isRead = false,
        modifiedAt = System.currentTimeMillis(),
        syncedAt = null,
        isDeleted = false,
    )

    private fun createCalendarEvent(
        id: String,
        startDay: String,
    ): CalendarEventEntity = CalendarEventEntity(
        id = id,
        eventType = "shift",
        eventTypeId = "shift-1",
        startDay = startDay,
        endDay = startDay,
        startTime = 480,
        endTime = 960,
        totalHours = 480,
        notes = null,
        modifiedAt = System.currentTimeMillis(),
        syncedAt = null,
        isDeleted = false,
    )
}

File: frontend/react-web/src/features/sync/services/notificationPurgeService.test.ts
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';

import { purgePastNotifications, getTodayDateString } from './notificationPurgeService';

/**
 * Property 3: Client purge identifies correct records
 * Validates: Requirements 2.1, 2.3, 2.7
 */

// Mock the db module
vi.mock('@/data/db', () => ({
  db: {
    notifications: {
      toArray: vi.fn(),
      bulkDelete: vi.fn(),
    },
    calendarEvents: {
      toArray: vi.fn(),
    },
  },
}));

import { db } from '@/data/db';

const mockNotificationsToArray = vi.mocked(db.notifications.toArray);
const mockCalendarEventsToArray = vi.mocked(db.calendarEvents.toArray);
const mockBulkDelete = vi.mocked(db.notifications.bulkDelete);

describe('notificationPurgeService', () => {
  beforeEach(() => {
    vi.clearAllMocks();
    mockBulkDelete.mockResolvedValue(undefined);
    // Fix today's date to 2025-06-15 for deterministic tests
    vi.useFakeTimers();
    vi.setSystemTime(new Date('2025-06-15T10:00:00'));
  });

  afterEach(() => {
    vi.useRealTimers();
  });

  describe('purgePastNotifications', () => {
    it('should return purgedCount: 0 when no notifications exist', async () => {
      mockNotificationsToArray.mockResolvedValue([]);

      const result = await purgePastNotifications();

      expect(result).toEqual({ purgedCount: 0 });
      expect(mockCalendarEventsToArray).not.toHaveBeenCalled();
      expect(mockBulkDelete).not.toHaveBeenCalled();
    });

    it('should purge notifications whose calendarEvent startDay is before today', async () => {
      mockNotificationsToArray.mockResolvedValue([
        { id: 'notif-1', calendarEventId: 'event-past' } as never,
        { id: 'notif-2', calendarEventId: 'event-past-2' } as never,
      ]);
      mockCalendarEventsToArray.mockResolvedValue([
        { id: 'event-past', startDay: '2025-06-10' } as never,
        { id: 'event-past-2', startDay: '2025-06-01' } as never,
      ]);

      const result = await purgePastNotifications();

      expect(result).toEqual({ purgedCount: 2 });
      expect(mockBulkDelete).toHaveBeenCalledWith(['notif-1', 'notif-2']);
    });

    it('should NOT purge notifications whose calendarEvent startDay equals today', async () => {
      mockNotificationsToArray.mockResolvedValue([
        { id: 'notif-today', calendarEventId: 'event-today' } as never,
      ]);
      mockCalendarEventsToArray.mockResolvedValue([
        { id: 'event-today', startDay: '2025-06-15' } as never,
      ]);

      const result = await purgePastNotifications();

      expect(result).toEqual({ purgedCount: 0 });
      expect(mockBulkDelete).not.toHaveBeenCalled();
    });

    it('should NOT purge notifications whose calendarEvent startDay is after today', async () => {
      mockNotificationsToArray.mockResolvedValue([
        { id: 'notif-future', calendarEventId: 'event-future' } as never,
      ]);
      mockCalendarEventsToArray.mockResolvedValue([
        { id: 'event-future', startDay: '2025-06-20' } as never,
      ]);

      const result = await purgePastNotifications();

      expect(result).toEqual({ purgedCount: 0 });
      expect(mockBulkDelete).not.toHaveBeenCalled();
    });

    it('should purge orphaned notifications (calendarEventId does not match any event)', async () => {
      mockNotificationsToArray.mockResolvedValue([
        { id: 'notif-orphan', calendarEventId: 'non-existent-event' } as never,
      ]);
      mockCalendarEventsToArray.mockResolvedValue([
        { id: 'some-other-event', startDay: '2025-06-20' } as never,
      ]);

      const result = await purgePastNotifications();

      expect(result).toEqual({ purgedCount: 1 });
      expect(mockBulkDelete).toHaveBeenCalledWith(['notif-orphan']);
    });

    it('should handle mixed records (some past, some current, some orphaned)', async () => {
      mockNotificationsToArray.mockResolvedValue([
        { id: 'notif-past', calendarEventId: 'event-past' } as never,
        { id: 'notif-today', calendarEventId: 'event-today' } as never,
        { id: 'notif-future', calendarEventId: 'event-future' } as never,
        { id: 'notif-orphan', calendarEventId: 'event-missing' } as never,
      ]);
      mockCalendarEventsToArray.mockResolvedValue([
        { id: 'event-past', startDay: '2025-06-10' } as never,
        { id: 'event-today', startDay: '2025-06-15' } as never,
        { id: 'event-future', startDay: '2025-06-20' } as never,
      ]);

      const result = await purgePastNotifications();

      expect(result).toEqual({ purgedCount: 2 });
      expect(mockBulkDelete).toHaveBeenCalledWith(['notif-past', 'notif-orphan']);
    });

    it('should return error in result (not throw) when IndexedDB operation fails', async () => {
      mockNotificationsToArray.mockRejectedValue(new Error('IndexedDB read failed'));

      const result = await purgePastNotifications();

      expect(result).toEqual({ purgedCount: 0, error: 'IndexedDB read failed' });
    });

    it('should call bulkDelete with correct IDs', async () => {
      mockNotificationsToArray.mockResolvedValue([
        { id: 'a1', calendarEventId: 'ev-old' } as never,
        { id: 'b2', calendarEventId: 'ev-current' } as never,
        { id: 'c3', calendarEventId: 'ev-orphan-ref' } as never,
      ]);
      mockCalendarEventsToArray.mockResolvedValue([
        { id: 'ev-old', startDay: '2025-05-01' } as never,
        { id: 'ev-current', startDay: '2025-06-15' } as never,
      ]);

      await purgePastNotifications();

      expect(mockBulkDelete).toHaveBeenCalledTimes(1);
      expect(mockBulkDelete).toHaveBeenCalledWith(['a1', 'c3']);
    });
  });

  describe('getTodayDateString', () => {
    it('should return a string in YYYY-MM-DD format', () => {
      const result = getTodayDateString();
      expect(result).toMatch(/^\d{4}-\d{2}-\d{2}$/);
    });

    it('should return the current date based on system time', () => {
      const result = getTodayDateString();
      expect(result).toBe('2025-06-15');
    });
  });
});
